File: vi/codes/kotlin/chapter_backtracking/permutations_i.kt
/**
 * File: permutations_i.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_backtracking.permutations_i

/* Giải thuật quay lui: Hoán vị I */
fun backtrack(
    state: MutableList<Int>,
    choices: IntArray,
    selected: BooleanArray,
    res: MutableList<MutableList<Int>?>
) {
    // Khi độ dài trạng thái bằng số lượng phần tử, ghi lại lời giải
    if (state.size == choices.size) {
        res.add(state.toMutableList())
        return
    }
    // Duyệt qua tất cả các lựa chọn
    for (i in choices.indices) {
        val choice = choices[i]
        // Cắt tỉa: không cho phép chọn lại phần tử đã chọn
        if (!selected[i]) {
            // Thử: thực hiện lựa chọn, cập nhật trạng thái
            selected[i] = true
            state.add(choice)
            // Tiến hành vòng lựa chọn tiếp theo
            backtrack(state, choices, selected, res)
            // Quay lui: hủy lựa chọn, khôi phục về trạng thái trước đó
            selected[i] = false
            state.removeAt(state.size - 1)
        }
    }
}

/* Hoán vị I */
fun permutationsI(nums: IntArray): MutableList<MutableList<Int>?> {
    val res = mutableListOf<MutableList<Int>?>()
    backtrack(mutableListOf(), nums, BooleanArray(nums.size), res)
    return res
}

/* Đoạn mã điều khiển */
fun main() {
    val nums = intArrayOf(1, 2, 3)

    val res = permutationsI(nums)

    println("Input array nums = ${nums.contentToString()}")
    println("All permutations res = $res")
}

File: vi/codes/go/chapter_backtracking/permutations_i.go
// File: permutations_i.go
// Created Time: 2023-05-14
// Author: Reanon (793584285@qq.com)

package chapter_backtracking

/* Giải thuật quay lui: hoán vị I */
func backtrackI(state *[]int, choices *[]int, selected *[]bool, res *[][]int) {
	// Khi độ dài state bằng số lượng phần tử, ghi lại lời giải
	if len(*state) == len(*choices) {
		newState := append([]int{}, *state...)
		*res = append(*res, newState)
	}
	// Duyệt qua tất cả các lựa chọn
	for i := 0; i < len(*choices); i++ {
		choice := (*choices)[i]
		// Cắt tỉa: không cho phép chọn lại một phần tử đã được chọn
		if !(*selected)[i] {
			// Thử: thực hiện lựa chọn, cập nhật state
			(*selected)[i] = true
			*state = append(*state, choice)
			// Tiến hành vòng lựa chọn tiếp theo
			backtrackI(state, choices, selected, res)
			// Quay lui: hoàn tác lựa chọn, khôi phục về state trước đó
			(*selected)[i] = false
			*state = (*state)[:len(*state)-1]
		}
	}
}

/* Hoán vị I */
func permutationsI(nums []int) [][]int {
	res := make([][]int, 0)
	state := make([]int, 0)
	selected := make([]bool, len(nums))
	backtrackI(&state, &nums, &selected, &res)
	return res
}
